### PR TITLE
NPSP Contact Merge bug fix and page updates

### LINKED ISSUE #5568: "See Duplicate Contact" option on NPSP Contact Merge shows error: Map key <DRS Record ID> not found in map
[Hub Post](https://powerofus.force.com/s/question/0D51E00005K6dpUSAR/error-when-trying-to-use-duplicate-contact-button-on-contact-merge)

**Issue:**
When a Contact is created that matches a Lead, a Duplicate Record Set (DRS) is created using a Lead duplicate rule. When the customer clicks on the See Duplicate Contacts button on the NPSP Contact Merge tab, the CON_ContactMerge_CTRL class has a method called showDuplicateRecordSets, which is failing. It's attempting to locate all the contacts related to the first contact, but it's failing as there are no contacts (there are Leads).
Screen Shot 2020-06-12 at 4 28 18 PM

Error:
Screen Shot 2020-06-12 at 4 29 57 PM

Full Error: 
*Map key <DRS ID> not found in map*
*Error is in expression '{!URLFOR($Action.Contact.View, firstContactOfDRS[drs.Id])}' in component <apex:repeat> in page npsp:con_contactmerge*

Seems to be failing in the showDuplicateRecordSets method, in the CON_ContactMerge_CTRL class, here (https://github.com/SalesforceFoundation/NPSP/blob/49dab6a0ff6d190ec131ed184d27870d7e49ee01/src/classes/CON_ContactMerge_CTRL.cls#L1025). 
This part (https://github.com/SalesforceFoundation/NPSP/blob/49dab6a0ff6d190ec131ed184d27870d7e49ee01/src/classes/CON_ContactMerge_CTRL.cls#L1056) is supposed to locate all contacts related to the first contact and display them per the point of the method, but it’s failing. I believe because there are no contacts, there are leads related to the first contact. 

**Steps to Reproduce:**
1. In an org with the NPSP, Duplicate Management, and the “Standard Rule for Contacts with Duplicate Leads” duplicate rule enabled, create a Lead. 
2. Create a matching Contact record with matching First/Last name as the Lead.
3. Confirm a new Duplicate Record Set has been automatically created and that it used the “Standard Rule for Contacts with Duplicate Leads” duplicate rule.
Screen Shot 2020-06-12 at 4 34 35 PM
4. Click on the Contact Merge tab, then the "See Duplicate Contacts" button. 

Expected Behaviour: Page is displayed showing no contacts to merge based on existing DRS records:
Screen Shot 2020-06-12 at 4 45 57 PM

Actual Behaviour: Error is displayed.

### PR DESCRIPTION
**Pull request by npsp-reedestockton**
# Critical Changes

# Changes
 - We made these updates to the See Duplicate Contacts page in Contact Merge:
   - Duplicate Record Sets that include a Lead Duplicate Record no longer causes an error.
   - The "Record Count" column has been renamed to "Contact Count".
   - Duplicate Record Sets that include one or more Leads and only one Contact are not displayed.
   - Duplicate Record Sets that include one or more Leads and two or more Contacts are displayed.
   - When you click the Pick button, only Contact records are displayed as merge candidates.

# Issues Closed
Fixes #5568 

# Community Ideas Delivered

# Features Intended for Future Release

# Features for Elevate Customers

# New Metadata

# Deleted Metadata

### CODE REVIEW

**Comment by npsp-reedestockton** [comment]
Tracks [W-7693961](https://gus.lightning.force.com/a07B0000008S1AGIA0)

**Comment by npsp-reedestockton** [comment]
@sam-knox Would appreciate your comments on PR description and title.  Also, there is a new label.  Replacing "Record Count" column with "Contact Count".  Not sure how creative you can be with that label change, but give it your best shot!

File: force-app/main/default/classes/CON_ContactMerge_CTRL.cls [hunk 1]
```diff
@@ -1041,71 +1053,97 @@ public with sharing class CON_ContactMerge_CTRL {
...
-                        firstContactOfDRS.get(dri.DuplicateRecordSetId).createdDate) {
-                        firstContactOfDRS.put(dri.DuplicateRecordSetId,con);
-                    }
+            Contact con = contacts.get(dri.RecordId);
+            if (!(firstContactOfDRS.containsKey(dri.DuplicateRecordSetId))) {
+                firstContactOfDRS.put(dri.DuplicateRecordSetId, con);
+            } else {
+                if (con.createdDate <
+                    firstContactOfDRS.get(dri.DuplicateRecordSetId).createdDate) {
+                    firstContactOfDRS.put(dri.DuplicateRecordSetId, con);
```
Comment (force2b) [inline force-app/main/default/classes/CON_ContactMerge_CTRL.cls:1113]: Is it worth abstracting out the contents of the for loop into its own method?

File: force-app/main/default/pages/CON_ContactMerge.page [hunk 2]
```diff
@@ -164,7 +164,7 @@
                                         <th class="slds-text-heading_label" scope="col">Duplicate Record Set Name</th>
                                         <th class="slds-text-heading_label" scope="col">Last Modified Date</th>
                                         <th class="slds-text-heading_label" scope="col">Duplicate Rule</th>
-                                        <th class="slds-text-heading_label" scope="col">Record Count</th>
+                                        <th class="slds-text-heading_label" scope="col">{!$Label.commonContactCount}</th>
```
Comment (force2b) [inline force-app/main/default/pages/CON_ContactMerge.page:167]: It looks like this page has a lot of hard-coded text (wow). Not that it all has to be fixed here, but can you open a new localization WI to fix the rest of the labels (if you haven't done that already)?

File: force-app/main/default/classes/CON_ContactMerge_CTRL.cls [hunk 3]
```diff
@@ -1041,71 +1053,97 @@ public with sharing class CON_ContactMerge_CTRL {
...
+        driFieldNames.add('Record.Type');
+        List<DuplicateRecordSet> duplicateRecordSets = new List<DuplicateRecordSet>(
+            (List<DuplicateRecordSet>) Database.query(constructDRSSOQL()));
+        driFieldNames.remove(driFieldNames.indexOf('DuplicateRecordSetId'));
+        driFieldNames.remove(driFieldNames.indexOf('Record.Type'));
+
+        return duplicateRecordSets;
+    }
+
+    private void getDRSFirstContacts(Set<Id> allDRIIds, List<DuplicateRecordItem> allDRIs) {
```
Comment (dfuller114) [inline force-app/main/default/classes/CON_ContactMerge_CTRL.cls:1092]: could we use duplicateRecordSet in the method name instead of abbreviation?

**Review by dfuller114 — CHANGES_REQUESTED** [review]
Looks good Reede! I've added a few comments regarding renaming abbreviated variables and methods. I think we should do as much of that for some of the legacy code in these classes as we can to clarify what the code is doing. I didn't add a comment for every occurrence, but I think we should at least remove abbreviations from method names and variables also if there's time. I've found the renaming Refactor feature in Illuminated Cloud to be really helpful for this.

**Comment by npsp-reedestockton** [comment]
@sam-knox Added another label on this one to replace a hard-coded field label - 'First Contact Name'

**Comment by npsp-reedestockton** [comment]
@force2b @dfuller114 Code review suggestions have been implemented.  Thanks y'all!

File: force-app/main/default/pages/CON_ContactMerge.page (hunk 2 above)
Comment (force2b) [reply force-app/main/default/pages/CON_ContactMerge.page:173]: Labels very nicely handled!

**Review by force2b — APPROVED** [review]

**Comment by sam-knox** [comment]
@npsp-reedestockton - Updated the title and change notes. Labels are good to go as they are!

**Review by dfuller114 — APPROVED** [review]
Looks good! Thanks Reede!

**Pull request merged**